src/engine/move_handling.h

```diff
@@ -71,7 +71,7 @@ constexpr void performPromotionMove(BitBoard& board, const movegen::Move& move,
 
     /* clear piece that will be taken if capture */
     if (move.isCapture()) {
-        if (const auto victim = board.getPieceAtSquare(move.toSquare())) {
+        if (const auto victim = (board.player == PlayerWhite) ? board.getPieceAtSquare<PlayerWhite>(move.toSquare()) : board.getPieceAtSquare<PlayerBlack>(move.toSquare())) {
```

Comment:
In the current implementation, we don't save on any checks for the colour, because the function calling `getPieceAtSquare` has not already checked for the colour.

It seems that if we want gains from the templating, we'd have to pass the pattern higher.

What do you think?

Reply:
I agree and added templating there, but also saw an opportunity for more significant optimization:

Profiling showed that the heaviest use of `getPieceAtSquare` was in `sortMoves`, where the latter is called for each valid move.

By templating `moveScore`, I've extracted the colour check from the 'for each move' loop, which should get us closer to a genuine half search. It only added a few lines to `sortMoves` so even if the instruction cache is a little bit more full, the gains should be a a net positive.

What do you think?